internal/venice-common/src/main/java/com/linkedin/venice/helix/StringAnnotatedHelixReadOnlySchemaRepository.java

```diff
@@ -0,0 +1,150 @@
+ */
+public class StringAnnotatedHelixReadOnlySchemaRepository implements ReadOnlySchemaRepository {
+  private final ReadOnlySchemaRepository internalSchemaRepo;
+  private final Map<String, Map<Integer, SchemaEntry>> valueSchemaEntryMapCache = new VeniceConcurrentHashMap<>();
+  private final Map<String, Map<String, DerivedSchemaEntry>> partialUpdateSchemaEntryMapCache =
+      new VeniceConcurrentHashMap<>();
+
+  public StringAnnotatedHelixReadOnlySchemaRepository(ReadOnlySchemaRepository internalSchemaRepo) {
+    this.internalSchemaRepo = internalSchemaRepo;
+  }
+
+  @Override
+  public void refresh() {
+    internalSchemaRepo.refresh();
+  }
+
+  @Override
+  public void clear() {
+    internalSchemaRepo.clear();
+  }
+
+  @Override
+  public SchemaEntry getKeySchema(String storeName) {
+    return internalSchemaRepo.getKeySchema(storeName);
+  }
+
+  @Override
+  /**
+   * Retrieve value schema of a store and annotate its map fields. The annotation will only be done once in the repository's
+   * lifetime as the result is cached.
+   */
+  public SchemaEntry getValueSchema(String storeName, int id) {
+    if (!(valueSchemaEntryMapCache.containsKey(storeName) && valueSchemaEntryMapCache.get(storeName).containsKey(id))) {
+      SchemaEntry schemaEntry = internalSchemaRepo.getValueSchema(storeName, id);
+      SchemaEntry annotatedSchemaEntry = getAnnotatedStringMapValueSchemaEntry(schemaEntry);
+      valueSchemaEntryMapCache.putIfAbsent(storeName, new VeniceConcurrentHashMap<>());
+      valueSchemaEntryMapCache.get(storeName).put(id, annotatedSchemaEntry);
+    }
+    return valueSchemaEntryMapCache.get(storeName).get(id);
+  }
+
+  @Override
+  public boolean hasValueSchema(String storeName, int id) {
+    return internalSchemaRepo.hasValueSchema(storeName, id);
+  }
+
+  @Override
+  public int getValueSchemaId(String storeName, String valueSchemaStr) {
+    return internalSchemaRepo.getValueSchemaId(storeName, valueSchemaStr);
+  }
+
+  @Override
+  public Collection<SchemaEntry> getValueSchemas(String storeName) {
+    return internalSchemaRepo.getValueSchemas(storeName);
+  }
+
+  @Override
+  /**
+   * Retrieve the superset schema (if exists) or the latest value schema of a store and annotate its map fields.
+   * The annotation will be done everytime when this method is invoked, as the superset schema or the latest value schema
+   * of the store can be changing from time to time.
+   */
+  public SchemaEntry getSupersetOrLatestValueSchema(String storeName) {
+    SchemaEntry schemaEntry = internalSchemaRepo.getSupersetOrLatestValueSchema(storeName);
+    return getAnnotatedStringMapValueSchemaEntry(schemaEntry);
+  }
+
+  @Override
+  /**
+   * Retrieve the superset schema (if exists) and annotate its map fields.
+   * The annotation will be done everytime when this method is invoked, as the superset schema of the store can be
+   * changing from time to time.
+   */
```

Comment:
Can we maintain a cache between the original schema and the annotated schema to avoid duplicate annotation calls?